### PR TITLE
Apply hotfix 20160419

### PR DESCRIPTION
**Pull request by mauritsvanrees**

### CODE REVIEW

File: plone/dexterity/content.py [hunk 1]
```diff
@@ -354,22 +354,20 @@ def _set__name__(self, value):
...
+        # Returns the item's globally unique id.
         return IUUID(self)
 
     @security.private
     def notifyModified(self):
-        """Update creators and modification_date.
-
-        This is called from CMFCatalogAware.reindexObject.
-        """
+        # Update creators and modification_date.
```
Comment (jensens) [inline plone/dexterity/content.py]: Does it get published even when ``@security.private`` decorated?

Comment (mauritsvanrees) [inline plone/dexterity/content.py]: No. So technically this change is unneeded.
But:
1. This sets a good example.
2. This avoids problems when someone later changes it to protected or public, here or in a subclass.
3. I don't want to think about it. There is no need for this to be publishable, so it should not get a doc string.
4. Current personal mode: kill all doc strings. ;-)

Comment (jensens) [inline plone/dexterity/content.py]: ad 1: this need to go into the styleguide then
ad 2: very good reason
ad 3: another good reason
ad 4: go go go! :+1:

Comment (malthe) [inline plone/dexterity/content.py]: Perhaps it's better to fix the [publishable object requirements](http://docs.zope.org/zope2/zdgbook/ObjectPublishing.html#publishable-object-requirements) – or declare a different security model on the class level.

It's always been very surprising that having a docstring changes object publication status.

**Comment by malthe** [comment]
Dropping the docstrings here seems very arbitrary given that the rest of the Zope/Plone codebase – as well as the rest of the Python community – uses and encourages docstrings.

This security issue was created because developers understand less and less of Zope 2 and its security model. That's only natural given that Plone tried very hard to embrace the ZTK and leave the Zope 2 legacy behind. But it's not an option (and it's not clear that it's even a good thing).

I think this pull request should be reworked to simply declare security explicitly. It's the easiest and most obvious thing to do.

**Comment by mauritsvanrees** [comment]
Sure, I am tempted to write a package called for example experimental.publishable. Make available a method/decorator security.declarePublishable that adds something like __methodname__publishable=True. Then change the publisher to ignore docstrings and instead look for this attribute.
But we would probably need to add a ton of these decorators in lots of packages or get lots of very real errors.
It may be the most obvious to do, but not the easiest, unless you have a lever idea about this.
In the long run: sure.

**Comment by malthe** [comment]
The problem is that a developer can forget to protect a method with a permission declaration.

If you look at the [original commit](c617b30315299f7227f7d1df8857806548289445) it's pretty clear that security hasn't been thought about at all. Lack of peer review seems to be the real issue here.

I don't think anything can be changed in ZPublisher to make the situation better. It's more of an awareness situation.

There is probably room for a tool that uses fuzzing to try and investigate which endpoints are available and map out which roles are able to access a given one.

**Comment by mauritsvanrees** [comment]
Of course the permission declarations are also needed for allowing access to some methods from inside a Zope 2 page template or a Python skin script.

**Comment by vangheem** [comment]
Another guy from the security team here... I'm okay with either approach. We've been screwed by the this "feature" of the publisher too many times. Anything that can bring more awareness/fix the situation is a +1 for me. We should provide an environment where is much harder to make screw ups like this. Publishing things with docstrings by default is insane.

**Comment by jensens** [comment]
I  think we have two areas where we can get in changes in different pace: Plone Code and Zope Code.

What we can do is to patch the publisher to do additional checks before it publishes something just because it has a docstring. 

- Provide a BBB checking mode which at least warns if something tries to be published just because of a docstring w/o security declarations. 

ideas then:
- Provide a whitelist of Zope core package that need this feature
- Provide more detailed lists (like dotted paths to methods)
- Decorate methods that need to be be published in Plone code (as kind of direct whitelisting).
- ...

**Comment by malthe** [comment]
I agree that it is insane.

I guess it needs to be clear that there is a different between Zope objects that do not implement a traverser and simple browser views which do:
```python
    def publishTraverse(self, request, name):
        if name == 'index.html':
            return self.index

        raise NotFound(self, name, request)
```
That is, that Zope objects are very closely tied to the (object) publisher. I mean it's the "o" in Zope.

**Comment by malthe** [comment]
@jensens – I took a look at ZPublisher and I think that adding this piece to the base request traverser makes some sense in a Plone context at least:
```diff
--- a/src/ZPublisher/BaseRequest.py
+++ b/src/ZPublisher/BaseRequest.py
@@ -15,6 +15,7 @@
 
 from urllib import quote as urllib_quote
+from types import MethodType 
 import xmlrpc
 
 from Acquisition import aq_base
 from Acquisition.interfaces import IAcquirer
@@ -128,6 +129,12 @@ class DefaultPublishTraverse(object):
                     request.response.setStatus(200)
                 except AttributeError:
                     pass
+                else:
+                    if isinstance(subobject, MethodType) and getRoles(object, name, subobject, None) is None:
+                        raise Forbidden(
+                            "The object at %s has no security declaration." %
+                            URL
+                        )
```

**Comment by mauritsvanrees** [comment]
See this comment by me from an earlier go a hacking the publisher:
https://github.com/plone/Products.CMFPlone/issues/1504#issuecomment-205277360

**Comment by malthe** [comment]
I think Plone can fix this concern by having an automated test that looks at every Plone-defined persistent object and makes sure that all methods with a docstring have a security declaration.
```python
def test_security_declarations(self):
    self.assert_class_security(self.class_under_test)
```

**Comment by mauritsvanrees** [comment]
Do you mean a test to put in say `Products/CMFPlone/tests/testSecurityAssertions.py` so that it gets run with `bin/test`?
Or a test that is run at Zope startup?

I don't know how to get a list of all such persistent objects programmatically.

Or do you mean in some zca hook during publishing? So we know which object is about to be published and then we do a check and in case of problems we either warn or raise an Exception, maybe depending on a policy?
A `zope.publisher.interfaces.browser.IBrowserPublisher` adapter could work. `DefaultPublishTraverse`, our trusty docstring checker, is such an adapter. An own adapter could call that one first, and then decide what to do with the result.

Note that checking if published methods with docstring have a security declaration is only part of the solution. This pull request removes the docstring from `addCreator` even though it has `@security.protected(permissions.ModifyPortalContent)`. I am removing this because there is no reason to allow access to the `addCreator` url, protected or not. And switching to private is not good (I guess) because then you cannot access `addCreator` from a Zope 2 template or a skin script.

I do wonder if making `addCreator` and other methods not publishable, means that access with XML-RPC is suffering. I am not using this myself, but I wonder if this is a use case we want to support. For me, choosing between better security and degraded XML-RPC support is easy: better security wins.

**Comment by mauritsvanrees** [comment]
Is there anything against merging this pull request and the others from https://github.com/plone/Products.CMFPlone/issues/1539?

Malthe seemed against, preferring a deeper solution to the docstring publishing problem. I think that is good for the long run, but for the short run we should do something now.
The pull requests are simply a merge of the security hotfix plus a few similar extra fixes. I can certainly remove the extra fixes if this makes it more acceptable. But the pure hotfix parts should definitely be included in the next release of this and other packages.

And repeating my main point from above, just for clarity: adding a security declaration to some of these methods is not enough: there are methods that have a security declaration and that need it, and that we still do not want to publish. Those security declarations are there because they are also checked when they are accessed from a Zope 2 page template or Python skin script.

**Comment by jensens** [comment]
I am +1 for merging this and in the long run to do the full fix in the 
publisher.


Am 03. Mai 2016 8:59:17 nachm. schrieb Maurits van Rees 
<notifications@github.com>:

**Comment by vangheem** [comment]
I'm also +1

**Comment by malthe** [comment]
If you don't want the methods published, put them in a view. It's a mistake
to have private, non-underscored methods on a published object.

**Comment by mauritsvanrees** [comment]
Sure, agreed. Though again, this is not only about methods that should be private.
But this is 15 year old code, from before the existence of browser views. Or in the dexterity case it is code that is meant to have the same methods as code from 15 years ago, because lots of other code expects this.

**Comment by mauritsvanrees** [comment]
I have merged this to 2.2.x (Plone 4.3) too.

**Pull request merged**